### PR TITLE
docs: start the work of snapping docs to the repo/docker

### PR DESCRIPTION
**Pull request by mattklein123**
Part of https://github.com/envoyproxy/envoy/issues/6361

### CODE REVIEW

**Comment by mattklein123** [comment]
@bndw @moderation PTAL and let me know what you think. I think the only thing remaining after this is to figure out how to deal with the Dockerfiles. I read a bit, and I think one can use env var substitution in those files in FROM clauses, so I think with a bit of doc work we can probably make it happen. We might need to move some of the MD sandbox instructions into sphinx though to make this happen.

@moderation any interest in looking at ^ since you have been working in that area recently?

**Comment by mattklein123** [comment]
cc @htuch I fixed one of your TODOs from a while ago.

File: docs/conf.py [hunk 1]
```diff
@@ -14,18 +14,46 @@
...
 
+# https://stackoverflow.com/questions/44761197/how-to-use-substitution-definitions-with-code-blocks
+class SubstitutionCodeBlock(CodeBlock):  # type: ignore
+  """
+  Similar to CodeBlock but replaces placeholders with variables.
+  """
+
+  def run(self):
+    """
+    Replace placeholders with given variables.
```
Comment (htuch) [inline docs/conf.py]: Environment variables?

File: docs/conf.py [hunk 2]
```diff
@@ -14,18 +14,46 @@
...
+    """
+    app = self.state.document.settings.env.app
+    new_content = []
+    self.content = self.content  # type: List[str]
```
Comment (htuch) [inline docs/conf.py]: I'm a bit confused by this line :) Is this self-assigning to establish some additional type safety?

Comment (mattklein123) [reply docs/conf.py]: Like a good developer, I just copied this code from stack overflow without looking at it. I will remove this line. :)

File: docs/conf.py (hunk 1 above)
Comment (mattklein123) [reply docs/conf.py]: I don't think they are really environment variables? I can make the comment a bit better.

**Comment by mattklein123** [comment]
@htuch updated

**Pull request merged**